### PR TITLE
SonarLint API

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java [hunk 1]
```diff
@@ -64,9 +64,22 @@
 
   /**
    * @since 5.5
+   * @deprecated since 6.0 replaced by {@link #getRuntimeApiVersion()}
    */
+  @Deprecated
   Version getSonarQubeVersion();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java [hunk 2]
```diff
@@ -145,8 +145,18 @@ public Version getSonarQubeVersion() {
...
+    return sqVersion.get();
+  }
+
+  @Override
+  public boolean isSonarLintRuntime() {
+    return sqVersion.isSonarlintRuntime();
+  }
+
+  public SensorContextTester setRuntime(Version version, boolean isSonarLint) {
+    this.sqVersion = new SonarQubeVersion(version, isSonarLint);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java [hunk 3]
```diff
@@ -85,18 +85,29 @@
...
+      this.runtimeApiVersion = runtimeApiVersion;
+      this.sonarlintRuntime = sonarlintRuntime;
     }
 
     /**
-     * Runtime version of SonarQube
+     * @deprecated since 6.0
      */
+    @Deprecated
     public Version getSonarQubeVersion() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/internal/RuntimeApiVersionFactory.java [hunk 4]
```diff
@@ -30,19 +30,19 @@
...
 
   private static final String FILE_PATH = "/sq-version.txt";
 
-  private SonarQubeVersionFactory() {
+  private RuntimeApiVersionFactory() {
     // prevents instantiation
   }
 
-  public static SonarQubeVersion create(System2 system) {
+  public static SonarQubeVersion create(System2 system, boolean isSonarLint) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/internal/RuntimeApiVersionFactory.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/internal/RuntimeApiVersionFactory.java [hunk 5]
```diff
@@ -30,19 +30,19 @@
...
     try {
       URL url = system.getResource(FILE_PATH);
       String versionInFile = Resources.toString(url, StandardCharsets.UTF_8);
-      return new SonarQubeVersion(Version.parse(versionInFile));
+      return new SonarQubeVersion(Version.parse(versionInFile), isSonarLint);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/internal/RuntimeApiVersionFactory.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java [hunk 6]
```diff
@@ -117,36 +115,17 @@
...
-  public SonarQubeVersion(Version version) {
-    requireNonNull(version);
-    this.version = version;
-  }
-
-  public Version get() {
-    return this.version;
-  }
+@Deprecated
+public class SonarQubeVersion extends RuntimeApiVersion {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues
* CRITICAL 9 critical
* MINOR 4 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL org.sonarsource.sonarqube:sonar-plugin-api: Class org.sonar.graph.StringEdge removed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL org.sonarsource.sonarqube:sonar-plugin-api: Class org.sonar.graph.StringEdgeFactory removed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [ExtensionProvider.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/ExtensionProvider.java): Removed org.sonar.api.Extension from the set of implemented interfaces [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [Plugin.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java): In method 'public Plugin$Context(org.sonar.api.utils.Version)' the number of arguments has changed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): Removed field V5_5 [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): Removed field V5_6 [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): In method 'public SonarQubeVersion(org.sonar.api.utils.Version)' the number of arguments has changed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SensorContext.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java): Method 'public org.sonar.api.utils.Version getRuntimeApiVersion()' has been added to an interface [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SensorContext.java](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java): Method 'public boolean isSonarLintRuntime()' has been added to an interface [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. MINOR [SensorContextTester.java#L90](https://github.com/SonarSource/sonarqube/blob/8380bd5dd27bde3ff2ba1d3ba0b5ffb55fa2047d/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java#L90): Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 16 issues
* CRITICAL 9 critical
* MINOR 4 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL org.sonarsource.sonarqube:sonar-plugin-api: Class org.sonar.graph.StringEdge removed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL org.sonarsource.sonarqube:sonar-plugin-api: Class org.sonar.graph.StringEdgeFactory removed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [ExtensionProvider.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/ExtensionProvider.java): Removed org.sonar.api.Extension from the set of implemented interfaces [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [Plugin.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java): In method 'public Plugin$Context(org.sonar.api.utils.Version)' the number of arguments has changed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): Removed field V5_5 [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): Removed field V5_6 [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SonarQubeVersion.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java): In method 'public SonarQubeVersion(org.sonar.api.utils.Version)' the number of arguments has changed [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SensorContext.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java): Method 'public org.sonar.api.utils.Version getRuntimeApiVersion()' has been added to an interface [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. CRITICAL [SensorContext.java](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/SensorContext.java): Method 'public boolean isSonarLintRuntime()' has been added to an interface [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=clirr%3Aclirr-api-break)
1. MINOR [SensorContextTester.java#L90](https://github.com/SonarSource/sonarqube/blob/03263455d6d8701940f634ee4fbeb3da6dbeec73/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java#L90): Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/RuntimeApiVersion.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+  /**
+   * Constant for version 5.6
+   */
+  public static final Version V5_6 = Version.create(5, 6);
+
+  private final Version version;
+  private final boolean sonarlint;
+
+  public RuntimeApiVersion(Version version, boolean sonarlint) {
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/RuntimeApiVersion.java]: This class mixes version and environment. It must be split or the class should be renamed. 
For the second option, I suggest `SonarRuntime` with methods 
* `getApiVersion()` 
* `getEnvironment()` which returns an enum. Indeed having `isSonarLintRuntime()` restricts to 2 environments (value `false` means SonarQube). That prevents future additions.

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 8]
```diff
@@ -696,6 +699,15 @@ public NewRule setTemplate(boolean template) {
       return this;
     }
 
+    /**
+     * Should this rule be enabled by default. For example in SonarLint standalone.
+     * @since 6.0
+     */
+    public NewRule setActivatedByDefault(boolean activatedByDefault) {
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: This information is not used in SonarQube environment ?

Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: Currently, no. Could be used later to replace "Sonar way" concept.

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* MINOR 4 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [SensorContextTester.java#L90](https://github.com/SonarSource/sonarqube/blob/ed953a9aef97bc3b85b7a8b2f9b1c174aaf870d7/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java#L90): Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by simonbrandhof** [comment]
as discussed, we need an interface SonarRuntime with methods `Version getApiVersion()`, `getProduct()` which returns an enum, and `getSonarQubeSide()` with returns an enum and fails if `getProduct() != SonarQube`.

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java [hunk 9]
```diff
@@ -142,11 +144,21 @@ public SensorContextTester setActiveRules(ActiveRules activeRules) {
...
+    return sqVersion.getApiVersion();
+  }
+
+  @Override
+  public SonarProduct getRuntimeProduct() {
+    return sqVersion.getProduct();
+  }
+
+  public SensorContextTester setRuntime(Version version, SonarProduct product, SonarQubeSide sonarQubeSide) {
+    this.sqVersion = new SonarQubeVersion(version, product, sonarQubeSide);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java [hunk 10]
```diff
@@ -85,18 +85,27 @@
...
+    public Context(SonarRuntime sonarRuntime) {
+      this.sonarRuntime = sonarRuntime;
     }
 
     /**
-     * Runtime version of SonarQube
+     * @deprecated since 6.0 use {@link #getRuntimeApiVersion()}
      */
+    @Deprecated
     public Version getSonarQubeVersion() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/Plugin.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/internal/SonarRuntimeFactory.java [hunk 11]
```diff
@@ -23,26 +23,29 @@
...
     // prevents instantiation
   }
 
-  public static SonarQubeVersion create(System2 system) {
+  public static SonarQubeVersion create(System2 system, SonarProduct product, @Nullable SonarQubeSide sonarQubeSide) {
     try {
       URL url = system.getResource(FILE_PATH);
       String versionInFile = Resources.toString(url, StandardCharsets.UTF_8);
-      return new SonarQubeVersion(Version.parse(versionInFile));
+      return new SonarQubeVersion(Version.parse(versionInFile), product, sonarQubeSide);
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/internal/SonarRuntimeFactory.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/internal/SonarRuntimeFactory.java [hunk 12]
```diff
@@ -23,26 +23,29 @@
...
 
   private static final String FILE_PATH = "/sq-version.txt";
 
-  private SonarQubeVersionFactory() {
+  private SonarRuntimeFactory() {
     // prevents instantiation
   }
 
-  public static SonarQubeVersion create(System2 system) {
+  public static SonarQubeVersion create(System2 system, SonarProduct product, @Nullable SonarQubeSide sonarQubeSide) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/internal/SonarRuntimeFactory.java]: MINOR Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java [hunk 13]
```diff
@@ -117,36 +116,21 @@
...
-   * Constant for version 5.5
-   */
-  public static final Version V5_5 = Version.create(5, 5);
-
-  /**
-   * Constant for version 5.6
-   */
-  public static final Version V5_6 = Version.create(5, 6);
+@Deprecated
+public class SonarQubeVersion extends SonarRuntime {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/SonarQubeVersion.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* MINOR 4 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [SensorContextTester.java#L92](https://github.com/SonarSource/sonarqube/blob/33047b12ee9de08fabd8054d1cef4eacdd3aefe4/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java#L92): Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by henryju** [comment]
@simonbrandhof I made the changes we discussed. I'm not really exposing SonarQube "side" for now. Do you think it should be exposed in Plugin.Context?

File: sonar-plugin-api/src/main/java/org/sonar/api/SonarRuntime.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+ *      &lt;/configuration&gt;
+ *    &lt;/plugin&gt;
+ *  &lt;/plugins&gt;
+ * &lt;/build&gt;
+ * </pre>
+ *
+ *
+ * @since 6.0
+ */
+@BatchSide
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/SonarRuntime.java]: MINOR Remove this use of "BatchSide"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* MINOR 5 minor
* INFO 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [SensorContextTester.java#L92](https://github.com/SonarSource/sonarqube/blob/e5818438a289fe522c87245d5bd18d71406f0e39/sonar-plugin-api/src/main/java/org/sonar/api/batch/sensor/internal/SensorContextTester.java#L92): Remove this use of "SonarQubeVersion"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Pull request merged**